File: src/common/common/main/java/settingdust/mod_sets/util/ServiceLoaderUtil.java
package settingdust.mod_sets.util;

import org.slf4j.Logger;
import settingdust.mod_sets.ModSets;

import java.util.*;

public final class ServiceLoaderUtil {
    public static Logger defaultLogger = ModSets.LOGGER;

    private ServiceLoaderUtil() {
    }

    public static <T> ServiceLoader<T> load(Class<T> clazz, ModuleLayer layer) {
        return ServiceLoader.load(layer, clazz);
    }

    public static <T> ServiceLoader<T> load(Class<T> clazz) {
        return ServiceLoader.load(clazz);
    }

    public static <T> ServiceLoader<T> load(Class<T> clazz, ClassLoader cl) {
        return ServiceLoader.load(clazz, cl);
    }

    public static <T> T findService(Class<T> clazz, ServiceLoader<T> serviceLoader) {
        return findService(clazz, serviceLoader, defaultLogger);
    }

    public static <T> T findService(Class<T> clazz, ServiceLoader<T> serviceLoader, Logger logger) {
        Iterator<T> it = findServices(clazz, serviceLoader, logger, true).iterator();
        if (it.hasNext()) {
            return it.next();
        }
        throw new NoSuchElementException("No service found for " + clazz);
    }

    public static <T> Iterable<T> findServices(Class<T> clazz, ModuleLayer layer) {
        return findServices(clazz, load(clazz, layer), defaultLogger, true);
    }

    public static <T> Iterable<T> findServices(Class<T> clazz, ServiceLoader<T> serviceLoader, boolean required) {
        return findServices(clazz, serviceLoader, defaultLogger, required);
    }

    public static <T> Iterable<T> findServices(
        Class<T> clazz,
        ServiceLoader<T> serviceLoader,
        Logger logger,
        boolean required
    ) {
        List<T> results = new ArrayList<>();
        String prefix = "[" + logger.getName() + "] ";
        Iterator<ServiceLoader.Provider<T>> iterator = serviceLoader.stream().iterator();
        List<Throwable> errors = new ArrayList<>();
        Optional<ServiceLoader.Provider<T>> current = findNext(iterator, errors);
        boolean found = false;

        while (current.isPresent()) {
            ServiceLoader.Provider<T> provider = current.get();
            String providerName = provider.type().getName();

            logger.debug("{}Loading {}", prefix, providerName);
            try {
                results.add(provider.get());
                found = true;
            } catch (Throwable t) {
                IllegalStateException e = new IllegalStateException(prefix + "Loading " + providerName + " failed", t);
                errors.add(e);
                logger.debug("", e);
            }

            current = findNext(iterator, errors);
        }

        if (!found && required) {
            IllegalStateException exception = new IllegalStateException("Load service of " + clazz + " failed");
            if (errors.isEmpty()) {
                exception.addSuppressed(new NoSuchElementException("Can't find service for " + clazz));
            }
            for (Throwable t : errors) {
                exception.addSuppressed(t);
            }
            throw exception;
        }

        return results;
    }

    public static <T> void loadServices(
        Class<T> clazz,
        ServiceLoader<T> serviceLoader,
        Logger logger,
        boolean required
    ) {
        for (T ignored : findServices(clazz, serviceLoader, logger, required)) {
            // Force loading
        }
    }

    private static <T> Optional<ServiceLoader.Provider<T>> findNext(
        Iterator<ServiceLoader.Provider<T>> iterator,
        List<Throwable> errors
    ) {
        ServiceLoader.Provider<T> current = null;
        while (true) {
            try {
                if (!iterator.hasNext()) {
                    return Optional.empty();
                }
                current = iterator.next();
                break;
            } catch (NoSuchElementException e) {
                return Optional.empty();
            } catch (Throwable t) {
                errors.add(t);
            }
        }
        return Optional.ofNullable(current);
    }

    // Convenience overloads
    public static <T> T findService(Class<T> clazz) {
        return findService(clazz, load(clazz), defaultLogger);
    }

    public static <T> Iterable<T> findServices(Class<T> clazz) {
        return findServices(clazz, load(clazz), defaultLogger, true);
    }

    public static <T> Iterable<T> findServices(Class<T> clazz, boolean required) {
        return findServices(clazz, load(clazz), defaultLogger, required);
    }

    public static <T> int loadServices(Class<T> clazz) {
        int count = 0;
        for (T ignored : findServices(clazz, load(clazz), defaultLogger, true)) {
            count++;
        }
        return count;
    }
}

File: src/common/main/kotlin/settingdust/mod_sets/util/ServiceLoaderUtil.kt
package settingdust.mod_sets.util

import org.apache.logging.log4j.Logger
import java.util.*

object ServiceLoaderUtil {
    lateinit var defaultLogger: Logger

    inline fun <reified T> load() = ServiceLoader.load(T::class.java)!!

    fun <T> findService(clazz: Class<T>, serviceLoader: ServiceLoader<T>, logger: Logger = defaultLogger) =
        findServices(clazz, serviceLoader, logger).first()

    inline fun <reified T> findService(serviceLoader: ServiceLoader<T> = load(), logger: Logger = defaultLogger) =
        findService(T::class.java, serviceLoader, logger)

    fun <T> findServices(
        clazz: Class<T>,
        serviceLoader: ServiceLoader<T>,
        logger: Logger = defaultLogger,
        required: Boolean = true
    ) = sequence {
        val prefix = String.format("[%s] ", logger.name)
        val iterator = serviceLoader.stream().iterator()
        val errors = mutableListOf<Throwable>()
        var current = findNext(iterator, errors)
        var found = false
        while (current.isSuccess) {
            val providerName: String = current.getOrThrow().type().getName()

            logger.debug("${prefix}Loading $providerName")

            try {
                yield(current.getOrThrow().get())
                found = true
            } catch (t: Throwable) {
                val e = IllegalStateException("${prefix}Loading $providerName failed", t)
                errors.add(e)
                logger.debug(e)
            }

            current = findNext<T>(iterator, errors)
        }
        if (found || !required) return@sequence
        val exception = IllegalStateException("Load service of $clazz failed")
        if (errors.isEmpty()) {
            exception.addSuppressed(NoSuchElementException("Can't find service for $clazz"))
        }
        errors.forEach(exception::addSuppressed)
        throw exception
    }

    inline fun <reified T> findServices(
        serviceLoader: ServiceLoader<T> = load(),
        logger: Logger = defaultLogger,
        required: Boolean = true
    ) = findServices(T::class.java, serviceLoader, logger, required)

    fun <T> loadServices(
        clazz: Class<T>,
        serviceLoader: ServiceLoader<T>,
        logger: Logger = defaultLogger,
        required: Boolean = true
    ) {
        findServices(clazz, serviceLoader, logger, required).forEach { _ -> }
    }

    inline fun <reified T> loadServices(
        serviceLoader: ServiceLoader<T> = load(),
        logger: Logger = defaultLogger,
        required: Boolean = true
    ) =
        findServices(T::class.java, serviceLoader, logger, required).count()

    private fun <T> findNext(
        iterator: Iterator<ServiceLoader.Provider<T>>,
        errors: MutableList<Throwable>
    ): Result<ServiceLoader.Provider<T>> {
        var current: ServiceLoader.Provider<T>? = null
        do {
            try {
                current = iterator.next()
            } catch (e: NoSuchElementException) {
                return Result.failure(e)
            } catch (t: Throwable) {
                errors.add(t)
            }
        } while (current == null)
        return Result.success(current)
    }
}